---
title: "combined_QTL"
author: "Weier Guo"
date: '2022-08-08'
output: html_document
---

### Introduction ###
This script aims for QTL analysis based on three genotype dataset: P.nigra SNPs, P.deltoides SNPs, relative dosage.
It is not a conventional QTL, so we have to run analysis based on customized model. 
Model:
**Phenotype ~ intercept + dosage_genotype + P.deltoides_genotype + P.nigra_genotype**
It can be started with: 
**Y ~ beta0 + beta1*gtDeltoides + beta2*gtNigra*Dosage + e**
**gtDeltoides: (-1,1)**
**gtNigra: (-1,1)**
**Dosage: (0,1,2...) represents (deletion, normal, duplication, ...)**
Later, it can add on other variables including interactions.

### load library and data ###
library
```{r}
library(tidyverse)
library(qqman)
library(ggplot2)
```

dataset
```{r}
# genotypes
nigra_gt <- read_tsv("../output/Pnigra_NewOrderGeno_pos_20220808.txt")
deltoides_gt <- read_tsv("../output/Pdeltoides_nigraMarker_20220801.txt")
dosage_gt <- read_tsv("../output/Dosage_nigraMarker_20220808.txt")
nigra_gt
deltoides_gt
dosage_gt
```


```{r}
# phenotypes
biomass <- read_tsv("../../Kmer/input/pheno_output_table_blup_20220428.txt")
leaf <- read_tsv("../../Kmer/input/pheno_leaf_20220527.txt")
vessel <- read_csv("../../Kmer/input/pheno_vessel_20220522.csv")
biomass
leaf
vessel
```

### Convert P.nigra and P.deltoides genotypes to (-1,1) ###
```{r}
nigra_gt_cv <- nigra_gt %>%
  select(-Old.Name) %>%
  str_replace(string = "1", pattern = )
```



