---
title: "combined_QTL"
author: "Weier Guo"
date: '2022-08-08'
output: html_document
---

### Introduction ###
This script aims for QTL analysis based on three genotype dataset: P.nigra SNPs, P.deltoides SNPs, relative dosage.
It is not a conventional QTL, so we have to run analysis based on customized model. 
Model:
**Phenotype ~ intercept + dosage_genotype + P.deltoides_genotype + P.nigra_genotype**
It can be started with: 
**Y ~ beta0 + beta1*gtDeltoides + beta2*gtNigra*Dosage + e**
**gtDeltoides: (-1,1)**
**gtNigra: (-1,1)**
**Dosage: (0,1,2...) represents (deletion, normal, duplication, ...)**
Later, it can add on other variables including interactions.

### load library and data ###
library
```{r}
library(tidyverse)
library(qqman)
library(ggplot2)
library(stats)
library(ggbeeswarm)
```

dataset
```{r}
# genotypes
nigra_gt <- read_tsv("../F1_450/genotype/Pnigra_F1_delGeno_imputed_conMk_20230108.txt")
deltoides_gt <- read_tsv("../F1_450/genotype/Pdeltoides_F1_delGeno_imputed_conMk_20230108.txt")
dosage_gt <- read_tsv("../F1_450/genotype/Dosage_F1_delGeno_imputed_conMk_20230108.txt")
nigra_gt
deltoides_gt
dosage_gt
```



```{r}
# phenotypes
biomass <- read_tsv("../../Kmer/input/pheno_output_table_blup_20220428.txt")
leaf <- read_tsv("../../Kmer/input/pheno_leaf_20220527.txt")
vessel <- read_csv("../../Kmer/input/pheno_vessel_20220522.csv")
biomass
leaf
vessel
```

```{r}
leaf.list <- read_csv("../input/Leaf_pheno_20220824.csv")
leaf.list

biomass.list <- read_csv("../input/Biomass_pheno_20230108.csv")
biomass.list
```

```{r}
leaf.new <- tibble(leaf$Old.Name)
for (i in 1:length(leaf.list$V1)) {
  pheno.name <- leaf.list$V1[i]
  indice <- which(colnames(leaf) == pheno.name)
  leaf.new <- cbind(leaf.new, leaf[,indice])
}
colnames(leaf.new)[1] <- "Old.Name"
leaf.new
```


```{r}
biomass.new <- tibble(biomass$Old.Name)
for (i in 1:length(biomass.list$V1)) {
  pheno.name <- biomass.list$V1[i]
  indice <- which(colnames(biomass) == pheno.name)
  biomass.new <- cbind(biomass.new, biomass[,indice])
}
colnames(biomass.new)[1] <- "Old.Name"
biomass.new
```



### dataset preparation ###
#### Replace deleted segments (NAs) in Pnigra genotype into 0 ####
This step is to set a tractable number for deleted segments in Pnigra, which can be easy handling in the statistical model. 
1. Locate the deleted segments in dosage_gt
```{r}
dosage_gt_matrix <- as.matrix(dosage_gt)
del_seg <- which(dosage_gt_matrix == "0",arr.ind = TRUE)
```

2. Transfer Pnigra genotype into matrix, prepare for replacement
```{r}
nigra_gt_matrix <- as.matrix(nigra_gt)
```

3. Based on del_seg, replace NAs in nigra_gt_matrix into 0.
```{r}
for (i in 1:nrow(del_seg)) {
  row <- del_seg[i,1]
  col <- del_seg[i,2]
  nigra_gt_matrix[row,col] <- "0"
}
```

4. Transform nigra_gt_matrix back to tibble
```{r}
nigra_gt_new <- as.data.frame(nigra_gt_matrix)
dat <- nigra_gt_new %>%
  select(-Old.Name) 
dat <- as.data.frame(sapply(dat, as.numeric))
nigra_gt_n <- dat %>%
  mutate(Old.Name = nigra_gt_new[,1]) %>%
  select(Old.Name, everything())
nigra_gt_n <- tibble(nigra_gt_n)
head(nigra_gt_n)
```

### combine genotypes and phenotypes into one bulk file ###
1. Modify each genotype marker name, prepare for distinguishing after merge the tables.
```{r}
# P.deltoides
newcolname <- c("Old.Name")
for (i in 2:ncol(deltoides_gt)) {
  newmarker <- paste(colnames(deltoides_gt)[i],"D",sep = "_")
  newcolname <- c(newcolname,newmarker)
}
deltoides.gt <- deltoides_gt
colnames(deltoides.gt) <- newcolname
head(deltoides.gt)
```

```{r}
# P.nigra
newcolname <- c("Old.Name")
for (i in 2:ncol(nigra_gt)) {
  newmarker <- paste(colnames(nigra_gt)[i],"N",sep = "_")
  newcolname <- c(newcolname,newmarker)
}
nigra.gt <- nigra_gt 
colnames(nigra.gt) <- newcolname
head(nigra.gt)
```

```{r}
# Dosage
newcolname <- c("Old.Name")
for (i in 2:ncol(dosage_gt)) {
  newmarker <- paste(colnames(dosage_gt)[i],"I",sep = "_")
  newcolname <- c(newcolname,newmarker)
}
dosage.gt <- dosage_gt
colnames(dosage.gt) <- newcolname
head(dosage.gt)
```
2. Combine three genotypes
```{r}
Del.Nig.Dos.geno <- deltoides.gt[,1]
for (i in 2:ncol(deltoides.gt)) {
  marker <- cbind(deltoides.gt[,i],nigra.gt[,i],dosage.gt[,i])
  Del.Nig.Dos.geno <- cbind(Del.Nig.Dos.geno, marker)
}
head(Del.Nig.Dos.geno)
```

3. add phenotypes
```{r}
biomass.Del.Nig.Dos <- left_join(Del.Nig.Dos.geno,biomass.new,by = "Old.Name")
biomass.Del.Nig.Dos 
```

4. remove phenotypes carrying NAs
```{r}
biomass.343 <- biomass.Del.Nig.Dos[,1523:1576]
biomass.343
biomass.keep <- c()

for (i in 1:54) {
  pheno <- biomass.343[,i]
  if (length(pheno[is.na(pheno)==TRUE]) <= 10) {
    biomass.keep <- c(biomass.keep, i)
  }
}

biomass.343.n <- biomass.343 %>%
  select(biomass.keep)
biomass.343.n

# save available biomass phenotypes
biomass.343.save <- cbind(biomass.Del.Nig.Dos$Old.Name, biomass.343.n)
colnames(biomass.343.save)[1] <- "Old.Name"
write_csv(biomass.343.save, "../input/Biomass_pheno_343_20230109.csv")
```



5. add new biomass phenotype to bulk geno-pheno file
```{r}
biomass.DelNigDos <- cbind(Del.Nig.Dos.geno, biomass.343.n)
biomass.DelNigDos
```

6. extract leaf morphology 343 lines phenotypes, save them
```{r}
leaf.Del.Nig.Dos
```


### create linear model ###
1. load library
```{r}
library(lmerTest)
library(lme4)
```


2. pickup the target phenotype
```{r,eval=FALSE}
leaf.Del.Nig.Dos <- leaf.Del.Nig.Dos %>%
  select(Old.Name, Blup_Transfo_Num_Indents_field_year2, everything())
leaf.Del.Nig.Dos
```

3. Recurrently run linear model for each marker
```{r}
n <- summary(lm(Blup_Transfo_Num_Indents_field_year2 ~ Chr01_3747646_4340087_D + Chr01_3747646_4340087_N + Chr01_3747646_4340087_I, data = leaf.Del.Nig.Dos))
n
```

### Recurrently generate all leaf morphology Manhattan plot and pvalue files ###
```{r}
for (j in 1523:length(leaf.Del.Nig.Dos)) {
  Pheno <- leaf.Del.Nig.Dos[,j]
  pheno.name <- colnames(leaf.Del.Nig.Dos)[j]
  geno <- seq(2,1523,3)
  model.pvalue <- tibble(Marker = c("Test"), 
                        Intercept = c(123), 
                        Deltoides = c(123), 
                        Nigra = c(123), 
                        Dosage = c(123),
                        Del_PctExp = c(123),
                        Nig_PctExp = c(123),
                        Dos_PctExp = c(123),
                        R_squared = c(123))

  for (i in geno) {
    colnames(leaf.Del.Nig.Dos)
    m <- lm(Pheno ~ leaf.Del.Nig.Dos[,i] + leaf.Del.Nig.Dos[,i+1] + leaf.Del.Nig.Dos[,i+2], data = leaf.Del.Nig.Dos)
    msum <- summary(m)
    # get variance explained by each variable
    maov <- anova(m)
    maovss <- maov$`Sum Sq`
    maovPEall <- maovss/sum(maovss)*100
    maovPE <- maovPEall[1:3]
    # get P-value for each variable
    pvalue <- msum$coefficients[,4]
    if (length(pvalue) < 4) {
      diff <- 4 - length(pvalue)
      for (j in 1:diff) {
        pvalue <- c(pvalue,"NA")
      }
    }
    # get adj. r-squared for model efficiency
    rsq <- msum$adj.r.squared * 100
    # arrange parameters into one sheet
    P.value <- c(colnames(deltoides_gt)[(i+1)/3+1],pvalue,maovPE,rsq)
    model.pvalue <- rbind(model.pvalue,P.value)
  }
  model.pvalue <- model.pvalue %>% 
    filter(Marker != "Test") %>%
    mutate(Del_P_adj = p.adjust(as.numeric(Deltoides), "BH")) %>%
    mutate(Nig_P_adj = p.adjust(as.numeric(Nigra), "BH")) %>%
    mutate(Dos_P_adj = p.adjust(as.numeric(Dosage), "BH")) %>%
    mutate(CHR = as.numeric(str_replace(string = Marker, pattern = "(...)(..)_(.*)_(.*)",replacement = "\\2"))) %>%
    mutate(BP = as.numeric(str_replace(string = Marker, pattern = "(...)(..)_(.*)_(.*)", replacement = "\\3")))
  
  # save pvalue file
  write_csv(model.pvalue, paste("../F1_450/DND/BH/sparsed/Pvalue/Leaf/DelGeno/", pheno.name, ".csv", sep = ""))
  
  # Manhattan plot for four variables
  # Deltoides
  Del.man <- model.pvalue %>%
    select(Marker, CHR, BP, Del_P_adj) %>%
    filter(Del_P_adj != "NA")
  manhattan(Del.man, snp = "Marker", p = "Del_P_adj", suggestiveline = -log10(0.05), genomewideline =-log10(0.01), main = paste("Leaf_Del", pheno.name, sep = "_"))
  dev.copy(png, paste("../F1_450/DND/BH/sparsed/Manhattan/DelGeno/Leaf_Del_", pheno.name, ".png", sep = ""), width = 500, height = 300)
  dev.off()
  # Nigra
  Nig.man <- model.pvalue %>%
    select(Marker, CHR, BP, Nig_P_adj) %>%
    filter(Nig_P_adj != "NA")
  manhattan(Nig.man, snp = "Marker", p = "Nig_P_adj", suggestiveline = -log10(0.05), genomewideline =-log10(0.01), main = paste("Leaf_Nig", pheno.name, sep = "_"))
  dev.copy(png, paste("../F1_450/DND/BH/sparsed/Manhattan/DelGeno/Leaf_Nig_", pheno.name, ".png", sep = ""), width = 500, height = 300)
  dev.off()
  # Dosage
  Dos.man <- model.pvalue %>%
    select(Marker, CHR, BP, Dos_P_adj) %>%
    filter(Dos_P_adj != "NA")
  manhattan(Dos.man, snp = "Marker", p = "Dos_P_adj", suggestiveline = -log10(0.05), genomewideline =-log10(0.01), main = paste("Leaf_Dos", pheno.name, sep = "_"))
  dev.copy(png, paste("../F1_450/DND/BH/sparsed/Manhattan/DelGeno/Leaf_Dos_", pheno.name, ".png", sep = ""), width = 500, height = 300)
  dev.off()
}

```




### Recurrently generate all biomass Manhattan plot and pvalue files ###
```{r}
for (j in 1523:length(biomass.DelNigDos)) {
  Pheno <- biomass.DelNigDos[,j]
  pheno.name <- colnames(biomass.DelNigDos)[j]
  geno <- seq(2,1523,3)
  model.pvalue <- tibble(Marker = c("Test"), 
                        Intercept = c(123), 
                        Deltoides = c(123), 
                        Nigra = c(123), 
                        Dosage = c(123),
                        R_squared = c(123))

  for (i in geno) {
    colnames(biomass.DelNigDos)
    m <- lm(Pheno ~ biomass.DelNigDos[,i] + biomass.DelNigDos[,i+1] + biomass.DelNigDos[,i+2], data = biomass.DelNigDos)
    msum <- summary(m)
    # get P-value for each variable
    pvalue <- msum$coefficients[,4]
    if (length(pvalue) < 4) {
      diff <- 4 - length(pvalue)
      for (j in 1:diff) {
        pvalue <- c(pvalue,"NA")
      }
    }
    # get adj. r-squared for model efficiency
    rsq <- msum$adj.r.squared * 100
    # arrange parameters into one sheet
    P.value <- c(colnames(deltoides_gt)[(i+1)/3+1],pvalue,rsq)
    model.pvalue <- rbind(model.pvalue,P.value)
  }
  model.pvalue <- model.pvalue %>% 
    filter(Marker != "Test") %>%
    mutate(Del_P_adj = p.adjust(as.numeric(Deltoides), "BH")) %>%
    mutate(Nig_P_adj = p.adjust(as.numeric(Nigra), "BH")) %>%
    mutate(Dos_P_adj = p.adjust(as.numeric(Dosage), "BH")) %>%
    mutate(CHR = as.numeric(str_replace(string = Marker, pattern = "(...)(..)_(.*)_(.*)",replacement = "\\2"))) %>%
    mutate(BP = as.numeric(str_replace(string = Marker, pattern = "(...)(..)_(.*)_(.*)", replacement = "\\3")))
  
  # save pvalue file
  write_csv(model.pvalue, paste("../F1_450/DND/BH/sparsed/Pvalue/Biomass/DelGeno/", pheno.name, ".csv", sep = ""))
  
  # Manhattan plot for four variables
  # Deltoides
  Del.man <- model.pvalue %>%
    select(Marker, CHR, BP, Del_P_adj) %>%
    filter(Del_P_adj != "NA")
  manhattan(Del.man, snp = "Marker", p = "Del_P_adj", suggestiveline = -log10(0.1), genomewideline =-log10(0.05), main = paste("Leaf_Del", pheno.name, sep = "_"))
  dev.copy(png, paste("../F1_450/DND/BH/sparsed/Manhattan/Biomass/DelGeno/Biomass_Del_", pheno.name, ".png", sep = ""), width = 500, height = 300)
  dev.off()
  # Nigra
  Nig.man <- model.pvalue %>%
    select(Marker, CHR, BP, Nig_P_adj) %>%
    filter(Nig_P_adj != "NA")
  manhattan(Nig.man, snp = "Marker", p = "Nig_P_adj", suggestiveline = -log10(0.1), genomewideline =-log10(0.05), main = paste("Leaf_Nig", pheno.name, sep = "_"))
  dev.copy(png, paste("../F1_450/DND/BH/sparsed/Manhattan/Biomass/DelGeno/Biomass_Nig_", pheno.name, ".png", sep = ""), width = 500, height = 300)
  dev.off()
  # Dosage
  Dos.man <- model.pvalue %>%
    select(Marker, CHR, BP, Dos_P_adj) %>%
    filter(Dos_P_adj != "NA")
  manhattan(Dos.man, snp = "Marker", p = "Dos_P_adj", suggestiveline = -log10(0.1), genomewideline =-log10(0.05), main = paste("Leaf_Dos", pheno.name, sep = "_"))
  dev.copy(png, paste("../F1_450/DND/BH/sparsed/Manhattan/Biomass/DelGeno/Biomass_Dos_", pheno.name, ".png", sep = ""), width = 500, height = 300)
  dev.off()
}

```

### model compare
```{r}
for (j in 1523:length(leaf.Del.Nig.Dos)) {
  Pheno <- leaf.Del.Nig.Dos[,j]
  pheno.name <- colnames(leaf.Del.Nig.Dos)[j]
  geno <- seq(2,1523,3)
  model.comp <- tibble(Marker = c("Test"), 
                        Del_Pval = c(123),
                        Nig_Pval = c(123),
                        Dos_Pval = c(123))

  for (i in geno) {
    colnames(leaf.Del.Nig.Dos)
    # set all models
    m.del <- lm(Pheno ~ leaf.Del.Nig.Dos[,i], data = leaf.Del.Nig.Dos)
    m.nig <- lm(Pheno ~ leaf.Del.Nig.Dos[,i+1], data = leaf.Del.Nig.Dos)
    m.dos <- lm(Pheno ~ leaf.Del.Nig.Dos[,i+2], data = leaf.Del.Nig.Dos)
    m.full <- lm(Pheno ~ leaf.Del.Nig.Dos[,i] + leaf.Del.Nig.Dos[,i+1] + leaf.Del.Nig.Dos[,i+2], data = leaf.Del.Nig.Dos)
    # get log-likelihood for each model
    m.del.loglik <- logLik(m.del)
    m.nig.loglik <- logLik(m.nig)
    m.dos.loglik <- logLik(m.dos)
    m.full.loglik <- logLik(m.full)
    # get test statistics
    m.del.stat <- -2 * (as.numeric(m.del.loglik)-as.numeric(m.full.loglik))
    m.nig.stat <- -2 * (as.numeric(m.nig.loglik)-as.numeric(m.full.loglik))
    m.dos.stat <- -2 * (as.numeric(m.dos.loglik)-as.numeric(m.full.loglik))
    # get p.values
    m.del.pval <- pchisq(m.del.stat, df=2, lower.tail = FALSE)
    m.nig.pval <- pchisq(m.nig.stat, df=2, lower.tail = FALSE)
    m.dos.pval <- pchisq(m.dos.stat, df=2, lower.tail = FALSE)
    
    # arrange parameters into one sheet
    model.p <- c(colnames(deltoides_gt)[(i+1)/3+1], m.del.pval, m.nig.pval, m.dos.pval)
    model.comp <- rbind(model.comp, model.p)
  }
  model.comp <- model.comp %>% 
    filter(Marker != "Test")
  
  # save model.comp file
  write_csv(model.comp, paste("../F1_450/model_comp/", pheno.name, ".csv", sep = ""))
}
```


```{r}
nig.sig1 %>%
  filter(is.na(Chr11_5977472_7452295_N) == FALSE) %>%
  ggplot(aes(x= Chr11_5977472_7452295_N, y=Blup_Transfo_PC1_PC2_y1_y2, group=Chr11_5977472_7452295_N)) +
  geom_violin(alpha=0.7) +
  geom_beeswarm(cex = 2, alpha=0.8) +
  xlab("P. nigra") +
  theme(text = element_text(size=15)) 
ggsave(filename = "../F1_450/plot/PC1_PC2_Pnigra_Chr11_5977472_7452295.png")

nig.sig1 %>%
  ggplot(aes(x=Chr11_5977472_7452295_I, y=Blup_Transfo_PC1_PC2_y1_y2, group=Chr11_5977472_7452295_I,color=Chr11_5977472_7452295_N)) +
  geom_violin(alpha=0.7) +
  geom_beeswarm(cex = 2, alpha=0.8) +
  xlab("Dosage") +
  theme(text = element_text(size=15)) +
  scale_color_discrete("P.nigra genotype")
ggsave(filename = "../F1_450/plot/PC1_PC2_Dosage_Chr11_5977472_7452295.png")

```

```{r}
nig.sig2 <- leaf.Del.Nig.Dos %>%
  select(Chr11_5977472_7452295_I, Chr11_5087792_5977448_N, Chr11_5977472_7452295_N, Chr11_7452456_8039698_N,Chr11_8039784_10539603_N, Blup_Transfo_PC1_PC2_y1_y2)

nig.sig2 %>%
  arrange(desc(Chr11_5977472_7452295_I))
```



### Recurrently run linear model with dosage as random effect for each marker
```{r}
Delt <- leaf.Del.Nig.Dos[,62]
Nigr <- leaf.Del.Nig.Dos[,63]
Dosa <- leaf.Del.Nig.Dos[,64]
m1 <- summary(lmer(leaf.Del.Nig.Dos[,2891] ~ Delt + Nigr * (1 | Dosa), data = leaf.Del.Nig.Dos))
m1
```

```{r}
for (j in 2891:length(leaf.Del.Nig.Dos)) {
  Pheno <- leaf.Del.Nig.Dos[,j]
  pheno.name <- colnames(leaf.Del.Nig.Dos)[j]
  geno <- seq(2,2890,3)
  lmer.pvalue <- tibble(Marker = c("Test"), 
                        Intercept = c(123), 
                        Deltoides = c(123), 
                        Nigra = c(123))

  for (i in geno) {
    colnames(leaf.Del.Nig.Dos)
    Del.gt <- leaf.Del.Nig.Dos[,i]
    Nig.gt <- leaf.Del.Nig.Dos[,i+1]
    Dos.gt <- leaf.Del.Nig.Dos[,i+2]
    m <- summary(lmer(Pheno ~ Del.gt + Nig.gt * (1 | Dos.gt), data = leaf.Del.Nig.Dos))
    pvalue <- m$coefficients[,5]
    if (length(pvalue) < 3) {
      diff <- 3 - length(pvalue)
      for (j in 1:diff) {
        pvalue <- c(pvalue,"NA")
      }
    }
    P.value <- c(colnames(deltoides_gt)[(i+1)/3+1],pvalue)
    lmer.pvalue <- rbind(lmer.pvalue,P.value)
  }
  lmer.pvalue <- lmer.pvalue %>% 
    filter(Marker != "Test") %>%
    mutate(Del_P_adj = p.adjust(as.numeric(Deltoides), "BH")) %>%
    mutate(Nig_P_adj = p.adjust(as.numeric(Nigra), "BH")) %>%
    mutate(CHR = as.numeric(str_replace(string = Marker, pattern = "(...)(..)_(.*)_(.*)",replacement = "\\2"))) %>%
    mutate(BP = as.numeric(str_replace(string = Marker, pattern = "(...)(..)_(.*)_(.*)", replacement = "\\3")))
  
  # save pvalue file
  write_csv(lmer.pvalue, paste("../lmer_pvalue_Leaf/", pheno.name, ".csv", sep = ""))
  
  # Manhattan plot for four variables
  # Deltoides
  Del.man <- lmer.pvalue %>%
    select(Marker, CHR, BP, Del_P_adj) %>%
    filter(Del_P_adj != "NA")
  manhattan(Del.man, snp = "Marker", p = "Del_P_adj", suggestiveline = -log10(0.05), genomewideline =-log10(0.1), main = paste("Manhattan_lmer_Del", pheno.name, sep = "_"))
  dev.copy(png, paste("../lmer_Manhattan_Leaf/Manhattan_Leaf/Man_lmer_Del_", pheno.name, ".png", sep = ""), width = 500, height = 300)
  dev.off()
  # Nigra
  Nig.man <- lmer.pvalue %>%
    select(Marker, CHR, BP, Nig_P_adj) %>%
    filter(Nig_P_adj != "NA")
  manhattan(Nig.man, snp = "Marker", p = "Nig_P_adj", suggestiveline = -log10(0.05), genomewideline =-log10(0.1), main = paste("Manhattan_lmer_Nig", pheno.name, sep = "_"))
  dev.copy(png, paste("../lm_Manhattan_Leaf/Man_lmer_Nig_", pheno.name, ".png", sep = ""), width = 500, height = 300)
  dev.off()
}

```





