---
title: "combined_QTL"
author: "Weier Guo"
date: '2022-08-08'
output: html_document
---

### Introduction ###
This script aims for QTL analysis based on three genotype dataset: P.nigra SNPs, P.deltoides SNPs, relative dosage.
It is not a conventional QTL, so we have to run analysis based on customized model. 
Model:
**Phenotype ~ intercept + dosage_genotype + P.deltoides_genotype + P.nigra_genotype**
It can be started with: 
**Y ~ beta0 + beta1*gtDeltoides + beta2*gtNigra*Dosage + e**
**gtDeltoides: (-1,1)**
**gtNigra: (-1,1)**
**Dosage: (0,1,2...) represents (deletion, normal, duplication, ...)**
Later, it can add on other variables including interactions.

### load library and data ###
library
```{r}
library(tidyverse)
library(qqman)
library(ggplot2)
```

dataset
```{r}
# genotypes
nigra_gt <- read_tsv("../output/Pnigra_NewOrderGeno_pos_20220812.txt")
deltoides_gt <- read_tsv("../output/Pdeltoides_nigraMarker_20220812.txt")
dosage_gt <- read_tsv("../output/Dosage_nigraMarker_20220812.txt")
nigra_gt
deltoides_gt
dosage_gt
```


```{r}
# phenotypes
biomass <- read_tsv("../../Kmer/input/pheno_output_table_blup_20220428.txt")
leaf <- read_tsv("../../Kmer/input/pheno_leaf_20220527.txt")
vessel <- read_csv("../../Kmer/input/pheno_vessel_20220522.csv")
biomass
leaf
vessel
```

```{r}
leaf.list <- read_csv("../input/Leaf_pheno_20220824.csv")
leaf.list
```

```{r}
leaf.new <- tibble(leaf$Old.Name)
for (i in 1:length(leaf.list$V1)) {
  pheno.name <- leaf.list$V1[i]
  indice <- which(colnames(leaf) == pheno.name)
  leaf.new <- cbind(leaf.new, leaf[,indice])
}
colnames(leaf.new)[1] <- "Old.Name"
leaf.new
```



### dataset preparation ###
#### Replace deleted segments (NAs) in Pnigra genotype into 0 ####
This step is to set a tractable number for deleted segments in Pnigra, which can be easy handling in the statistical model. 
1. Locate the deleted segments in dosage_gt
```{r}
dosage_gt_matrix <- as.matrix(dosage_gt)
del_seg <- which(dosage_gt_matrix == "0",arr.ind = TRUE)
```

2. Transfer Pnigra genotype into matrix, prepare for replacement
```{r}
nigra_gt_matrix <- as.matrix(nigra_gt)
```

3. Based on del_seg, replace NAs in nigra_gt_matrix into 0.
```{r}
for (i in 1:nrow(del_seg)) {
  row <- del_seg[i,1]
  col <- del_seg[i,2]
  nigra_gt_matrix[row,col] <- "0"
}
```

4. Transform nigra_gt_matrix back to tibble
```{r}
nigra_gt_new <- as.data.frame(nigra_gt_matrix)
dat <- nigra_gt_new %>%
  select(-Old.Name) 
dat <- as.data.frame(sapply(dat, as.numeric))
nigra_gt_n <- dat %>%
  mutate(Old.Name = nigra_gt_new[,1]) %>%
  select(Old.Name, everything())
nigra_gt_n <- tibble(nigra_gt_n)
head(nigra_gt_n)
```

### combine genotypes and phenotypes into one bulk file ###
1. Modify each genotype marker name, prepare for distinguishing after merge the tables.
```{r}
# P.deltoides
newcolname <- c("Old.Name")
for (i in 2:ncol(deltoides_gt)) {
  newmarker <- paste(colnames(deltoides_gt)[i],"D",sep = "_")
  newcolname <- c(newcolname,newmarker)
}
deltoides.gt <- deltoides_gt
colnames(deltoides.gt) <- newcolname
head(deltoides.gt)
```

```{r}
# P.nigra
newcolname <- c("Old.Name")
for (i in 2:ncol(nigra_gt_n)) {
  newmarker <- paste(colnames(nigra_gt_n)[i],"N",sep = "_")
  newcolname <- c(newcolname,newmarker)
}
nigra.gt <- nigra_gt_n 
colnames(nigra.gt) <- newcolname
head(nigra.gt)
```

```{r}
# Dosage
newcolname <- c("Old.Name")
for (i in 2:ncol(dosage_gt)) {
  newmarker <- paste(colnames(dosage_gt)[i],"I",sep = "_")
  newcolname <- c(newcolname,newmarker)
}
dosage.gt <- dosage_gt
colnames(dosage.gt) <- newcolname
head(dosage.gt)
```
2. Combine three genotypes
```{r}
Del.Nig.Dos.geno <- deltoides.gt[,1]
for (i in 2:ncol(deltoides.gt)) {
  marker <- cbind(deltoides.gt[,i],nigra.gt[,i],dosage.gt[,i])
  Del.Nig.Dos.geno <- cbind(Del.Nig.Dos.geno, marker)
}
head(Del.Nig.Dos.geno)
```

3. add phenotypes
```{r}
leaf.Del.Nig.Dos <- left_join(Del.Nig.Dos.geno,leaf.new,by = "Old.Name") 
leaf.Del.Nig.Dos
```

### create linear model ###
1. load library
```{r}
library(lmerTest)
```


2. pickup the target phenotype
```{r,eval=FALSE}
leaf.Del.Nig.Dos <- leaf.Del.Nig.Dos %>%
  select(Old.Name, Blup_Transfo_Num_Indents_field_year2, everything())
leaf.Del.Nig.Dos
```

3. Recurrently run linear model for each marker
```{r}
summary(lm(Blup_Transfo_Num_Indents_field_year2 ~ Chr01_126238_809425_D + Chr01_126238_809425_N * Chr01_126238_809425_I, data = leaf.Del.Nig.Dos))
```


```{r}
for (j in 1418:length(leaf.Del.Nig.Dos)) {
  Pheno <- leaf.Del.Nig.Dos[,j]
  pheno.name <- colnames(leaf.Del.Nig.Dos)[j]
  geno <- seq(2,1417,3)
  model.pvalue <- tibble(Marker = c("Test"), 
                        Intercept = c(123), 
                        Deltoides = c(123), 
                        Nigra = c(123), 
                        Dosage = c(123), 
                        Nigra_Dosage = c(123))

  for (i in geno) {
    colnames(leaf.Del.Nig.Dos)
    m <- summary(lm(Pheno ~ leaf.Del.Nig.Dos[,i] + leaf.Del.Nig.Dos[,i+1] * leaf.Del.Nig.Dos[,i+2], data = leaf.Del.Nig.Dos))
    pvalue <- m$coefficients[,4]
    if (length(pvalue) < 5) {
      diff <- 5 - length(pvalue)
      for (j in 1:diff) {
        pvalue <- c(pvalue,"NA")
      }
    }
    P.value <- c(colnames(deltoides_gt)[(i+1)/3+1],pvalue)
    model.pvalue <- rbind(model.pvalue,P.value)
  }
  model.pvalue <- model.pvalue %>% 
    filter(Marker != "Test") %>%
    mutate(Del_P_adj = p.adjust(as.numeric(Deltoides), "BH")) %>%
    mutate(Nig_P_adj = p.adjust(as.numeric(Nigra), "BH")) %>%
    mutate(Dos_P_adj = p.adjust(as.numeric(Dosage), "BH")) %>%
    mutate(NigDos_P_adj = p.adjust(as.numeric(Nigra_Dosage), "BH")) %>%
    mutate(CHR = as.numeric(str_replace(string = Marker, pattern = "(...)(..)_(.*)_(.*)",replacement = "\\2"))) %>%
    mutate(BP = as.numeric(str_replace(string = Marker, pattern = "(...)(..)_(.*)_(.*)", replacement = "\\3")))
  
  # Manhattan plot for four variables
  # Deltoides
  Del.man <- model.pvalue %>%
    select(Marker, CHR, BP, Del_P_adj) %>%
    filter(Del_P_adj != "NA")
  manhattan(Del.man, snp = "Marker", p = "Del_P_adj", suggestiveline = -log10(0.05), genomewideline =-log10(0.1), main = paste("Manhattan_Deltoides", pheno.name, sep = "_"))
  dev.copy(png, paste("Manhattan_Deltoides_", pheno.name, ".png", sep = ""), width = 500, height = 300)
  dev.off()
  # Nigra
  Nig.man <- model.pvalue %>%
    select(Marker, CHR, BP, Nig_P_adj) %>%
    filter(Nig_P_adj != "NA")
  manhattan(Nig.man, snp = "Marker", p = "Nig_P_adj", suggestiveline = -log10(0.05), genomewideline =-log10(0.1), main = paste("Manhattan_Nigra", pheno.name, sep = "_"))
  dev.copy(png, paste("Manhattan_Nigra_", pheno.name, ".png", sep = ""), width = 500, height = 300)
  dev.off()
  # Dosage
  Dos.man <- model.pvalue %>%
    select(Marker, CHR, BP, Dos_P_adj) %>%
    filter(Dos_P_adj != "NA")
  manhattan(Dos.man, snp = "Marker", p = "Dos_P_adj", suggestiveline = -log10(0.05), genomewideline =-log10(0.1), main = paste("Manhattan_Dosage", pheno.name, sep = "_"))
  dev.copy(png, paste("Manhattan_Dosage_", pheno.name, ".png", sep = ""), width = 500, height = 300)
  dev.off()
  # Nigra*Dosage
  NigDos.man <- model.pvalue %>%
    select(Marker, CHR, BP, NigDos_P_adj) %>%
    filter(NigDos_P_adj != "NA")
  manhattan(NigDos.man, snp = "Marker", p = "NigDos_P_adj", suggestiveline = -log10(0.05), genomewideline =-log10(0.1), main = paste("Manhattan_Nigra_Dosage", pheno.name, sep = "_"))
  dev.copy(png, paste("Manhattan_Nigra_Dosage_", pheno.name, ".png", sep = ""), width = 500, height = 300)
  dev.off()
}

```

### manhattan plot ###
```{r}
Del.man <- model.pvalue %>%
  select(Marker,CHR,BP,Del_P_adj) %>%
  filter(Del_P_adj != "NA")
manhattan(Del.man, snp = "Marker", p = "Del_P_adj", suggestiveline = -log10(0.05), genomewideline =-log10(0.1), main = "Manhattan_Deltoides_Blup_Transfo_Num_Indents_field_year2")
```

```{r}
Nig.man <- model.pvalue %>%
  select(Marker,CHR,BP,Nig_P_adj) %>%
  filter(Nig_P_adj != "NA")
manhattan(Nig.man, snp = "Marker", p = "Nig_P_adj", suggestiveline = -log10(0.05), genomewideline =-log10(0.1), main = "Manhattan_Nigra_Blup_Transfo_Num_Indents_field_year2")
```

```{r}
Dos.man <- model.pvalue %>%
  select(Marker,CHR,BP,Dos_P_adj) %>%
  filter(Dos_P_adj != "NA")
manhattan(Dos.man, snp = "Marker", p = "Dos_P_adj", suggestiveline = -log10(0.05), genomewideline =-log10(0.1), main = "Manhattan_Dosage_Blup_Transfo_Num_Indents_field_year2")
```

```{r}
NigDos.man <- model.pvalue %>%
  select(Marker,CHR,BP,NigDos_P_adj) %>%
  filter(NigDos_P_adj != "NA")
manhattan(NigDos.man, snp = "Marker", p = "NigDos_P_adj", suggestiveline = -log10(0.05), genomewideline =-log10(0.1), main = "Manhattan_Nigra_Dosage_Blup_Transfo_Num_Indents_field_year2")
```


### Select significant rows (pvalue < 0.05) ###
```{r}
# P.deltoides
del.sig <- model.pvalue %>%
  filter(Deltoides < 0.05) %>%
  mutate(Phenotype = "Num_Indent_y2") %>%
  select(Phenotype, everything())
del.sig
```

```{r,eval=FALSE}
# P.nigra
nig.sig <- model.pvalue %>%
  filter(Nigra < 0.05) %>%
  mutate(Phenotype = "Num_Indent_y2") %>%
  select(Phenotype, everything())
nig.sig
```

```{r,eval=FALSE}
# Dosage
dos.sig <- model.pvalue %>%
  filter(Dosage < 0.05) %>%
  mutate(Phenotype = "Num_Indent_y2") %>%
  select(Phenotype, everything())
dos.sig
```

```{r,eval=FALSE}
# Nigra*Dosage
nig.dos.sig <- model.pvalue %>%
  filter(Nigra_Dosage < 0.05) %>%
  mutate(Phenotype = "Num_Indent_y2") %>%
  select(Phenotype, everything())
nig.dos.sig
```

### merge phenotypes pvalue tables ###
```{r,eval=FALSE}
# P.deltoides
del.sig.add <- model.pvalue.add %>%
  filter(Deltoides < 0.05) %>%
  mutate(Phenotype = "PC3_PC4_y2") %>%
  select(Phenotype, everything())
del.sig <- rbind(del.sig, del.sig.add)
del.sig
```

```{r,eval=FALSE}
# P.nigra
nig.sig.add <- model.pvalue.add %>%
  filter(Nigra < 0.05) %>%
  mutate(Phenotype = "PC3_PC4_y2") %>%
  select(Phenotype, everything())
nig.sig <- rbind(nig.sig, nig.sig.add)
nig.sig
```

```{r,eval=FALSE}
# Dosage
dos.sig.add <- model.pvalue.add %>%
  filter(Dosage < 0.05) %>%
  mutate(Phenotype = "PC3_PC4_y2") %>%
  select(Phenotype, everything())
dos.sig <- rbind(dos.sig, dos.sig.add)
dos.sig
```

```{r,eval=FALSE}
# P.nigra*Dosage
nig.dos.sig.add <- model.pvalue.add %>%
  filter(Nigra_Dosage < 0.05) %>%
  mutate(Phenotype = "PC3_PC4_y2") %>%
  select(Phenotype, everything())
nig.dos.sig <- rbind(nig.dos.sig, nig.dos.sig.add)
nig.dos.sig
```

### output pvalue tables ###
```{r,eval=FALSE}
write_csv(del.sig, "../output/Leaf_deltoides_pvalue_20220822.csv")
write_csv(nig.sig, "../output/Leaf_nigra_pvalue_20220822.csv")
write_csv(dos.sig, "../output/Leaf_dosage_pvalue_20220822.csv")
write_csv(nig.dos.sig, "../output/Leaf_nigdos_pvalue_20220822.csv")
```


