---
title: "lmer_QTL"
author: "Weier Guo"
date: '2022-08-29'
output: html_document
---

### Introduction ###
A Notebook to run QTL with only P.deltoides and P.nigra. The effect of dosage variation should be embedded into P.nigra before the model establishment. 

Model:
**Phenotype ~ intercept + dosage_genotype + P.deltoides_genotype + P.nigra_genotype**
It can be started with: 
**Y ~ beta0 + beta1*gtDeltoides + beta2*gtNigra + e**
**gtDeltoides: (-1,1)**
**gtNigra: (-1,1,0,-2,2...) represents (allele1 normal, allele2 normal, deletion, allele1 duplication, allele2 duplication, ...)**


### load library and data ###
library
```{r}
library(tidyverse)
library(qqman)
library(ggplot2)
```

dataset
```{r}
# genotypes
nigra_gt <- read_tsv("../output/Pnigra_genome_marker_20220825.txt")
deltoides_gt <- read_tsv("../output/Pdeltoides_nigraMarker_20220825.txt")
dosage_gt <- read_tsv("../output/Dosage_nigraMarker_20220825.txt")
nigra_gt
deltoides_gt
dosage_gt
```

```{r}
# phenotypes
biomass <- read_tsv("../../Kmer/input/pheno_output_table_blup_20220428.txt")
leaf <- read_tsv("../../Kmer/input/pheno_leaf_20220527.txt")
vessel <- read_csv("../../Kmer/input/pheno_vessel_20220522.csv")
biomass
leaf
vessel
```


```{r}
leaf.list <- read_csv("../input/Leaf_pheno_20220824.csv")
leaf.list
```

```{r}
leaf.new <- tibble(leaf$Old.Name)
for (i in 1:length(leaf.list$V1)) {
  pheno.name <- leaf.list$V1[i]
  indice <- which(colnames(leaf) == pheno.name)
  leaf.new <- cbind(leaf.new, leaf[,indice])
}
colnames(leaf.new)[1] <- "Old.Name"
leaf.new
```


### dataset preparation ###
#### Embed dosage variation into P.nigra genotype ####
1. Transfer dosage genotype into matrix
```{r}
dosage_gt_matrix <- as.matrix(dosage_gt)
del_seg <- which(dosage_gt_matrix == "0",arr.ind = TRUE)
dup_seg <- which(dosage_gt_matrix == "2",arr.ind = TRUE)
```

2. Transfer Pnigra genotype into matrix, prepare for replacement
```{r}
nigra_gt_matrix <- as.matrix(nigra_gt)
```

3. Based on del_seg, replace NAs in nigra_gt_matrix into 0
```{r}
for (i in 1:nrow(del_seg)) {
  row <- del_seg[i,1]
  col <- del_seg[i,2]
  nigra_gt_matrix[row,col] <- "0"
}
```

4. Based on dup_seg, replace alleles in nigra_gt_matrix into -2 or 2 or keep in NAs. 
```{r}
for (i in 1:nrow(dup_seg)) {
  row <- dup_seg[i,1]
  col <- dup_seg[i,2]
  if (is.na(nigra_gt_matrix[row,col] == TRUE )) {
    nigra_gt_matrix[row,col] = "NA"
  }
  else {
    nigra_gt_matrix[row,col] <- as.character(as.numeric(nigra_gt_matrix[row,col]) * 2)
  }
}
```


5. Transform nigra_gt_matrix back to tibble
```{r}
nigra_gt_new <- as.data.frame(nigra_gt_matrix)
dat <- nigra_gt_new %>%
  select(-Old.Name) 
dat <- as.data.frame(sapply(dat, as.numeric))
nigra_gt_n <- dat %>%
  mutate(Old.Name = nigra_gt_new[,1]) %>%
  select(Old.Name, everything())
nigra_gt_n <- tibble(nigra_gt_n)
nigra_gt_n
```

### combine genotypes and phenotypes into one bulk file ###
1. Modify each genotype marker name, prepare for distinguishing after merge the tables.
```{r}
# P.deltoides
newcolname <- c("Old.Name")
for (i in 2:ncol(deltoides_gt)) {
  newmarker <- paste(colnames(deltoides_gt)[i],"D",sep = "_")
  newcolname <- c(newcolname,newmarker)
}
deltoides.gt <- deltoides_gt
colnames(deltoides.gt) <- newcolname
head(deltoides.gt)
```

```{r}
# P.nigra
newcolname <- c("Old.Name")
for (i in 2:ncol(nigra_gt_n)) {
  newmarker <- paste(colnames(nigra_gt_n)[i],"N",sep = "_")
  newcolname <- c(newcolname,newmarker)
}
nigra.gt <- nigra_gt_n 
colnames(nigra.gt) <- newcolname
head(nigra.gt)
```

```{r}
# Dosage
newcolname <- c("Old.Name")
for (i in 2:ncol(dosage_gt)) {
  newmarker <- paste(colnames(dosage_gt)[i],"I",sep = "_")
  newcolname <- c(newcolname,newmarker)
}
dosage.gt <- dosage_gt
colnames(dosage.gt) <- newcolname
head(dosage.gt)
```


2. Combine three genotypes
```{r}
Del.Nig.Dos.geno <- deltoides.gt[,1]
for (i in 2:ncol(deltoides.gt)) {
  marker <- cbind(deltoides.gt[,i],nigra.gt[,i],dosage.gt[,i])
  Del.Nig.Dos.geno <- cbind(Del.Nig.Dos.geno, marker)
}
head(Del.Nig.Dos.geno)
```

3. add phenotypes
```{r}
leaf.Del.Nig.Dos <- left_join(Del.Nig.Dos.geno,leaf.new,by = "Old.Name")
leaf.Del.Nig.Dos
```


### adjusted QTL ###
This step is to remove samples with dosage variations on chr02 markers, to identify whether there are other QTLs jumped out. 

1. Get the dosage variation samples on dosage_gt
```{r}
dosage.gt %>%
  filter(Chr02_4204964_5059530_I == "0")
```

2. remove these samples within three genotype lists 
```{r}
deltoides.gt.noC2 <- deltoides.gt %>%
  filter(Old.Name != "GWR_100_152" & Old.Name != "GWR_100_459")
nigra.gt.noC2 <- nigra.gt %>%
  filter(Old.Name != "GWR_100_152" & Old.Name != "GWR_100_459")
dosage.gt.noC2 <- dosage.gt %>%
  filter(Old.Name != "GWR_100_152" & Old.Name != "GWR_100_459")
```

3. combine three genotype lists
```{r}
Del.Nig.Dos.geno.noC2 <- deltoides.gt.noC2[,1]
for (i in 2:ncol(deltoides.gt.noC2)) {
  marker <- cbind(deltoides.gt.noC2[,i],nigra.gt.noC2[,i])
  Del.Nig.Dos.geno.noC2 <- cbind(Del.Nig.Dos.geno.noC2, marker)
}
head(Del.Nig.Dos.geno.noC2)
```

4. add phenotypes
```{r}
leaf.Del.Nig.Dos.noC2 <- left_join(Del.Nig.Dos.geno.noC2,leaf.new,by = "Old.Name")
leaf.Del.Nig.Dos.noC2
```



### create linear model ###
1. load library
```{r}
library(lmerTest)
library(lme4)
```

2. recurrently get Pvalue list and Manhattan plot 
```{r}
# Y ~ Deltoides + Nigra (embedded with dosage variation)
for (j in 1928:length(leaf.Del.Nig.Dos.noC2)) {
  Pheno <- leaf.Del.Nig.Dos.noC2[,j]
  pheno.name <- colnames(leaf.Del.Nig.Dos.noC2)[j]
  geno <- seq(2,1927,2)
  model.pvalue <- tibble(Marker = c("Test"), 
                        Intercept = c(123), 
                        Deltoides = c(123), 
                        Nigra = c(123))

  for (i in geno) {
    colnames(leaf.Del.Nig.Dos.noC2)
    m <- summary(lm(Pheno ~ leaf.Del.Nig.Dos.noC2[,i] + leaf.Del.Nig.Dos.noC2[,i+1], data = leaf.Del.Nig.Dos.noC2))
    pvalue <- m$coefficients[,4]
    if (length(pvalue) < 3) {
      diff <- 3 - length(pvalue)
      for (j in 1:diff) {
        pvalue <- c(pvalue,"NA")
      }
    }
    P.value <- c(colnames(deltoides_gt)[i/2+1],pvalue)
    model.pvalue <- rbind(model.pvalue,P.value)
  }
  model.pvalue <- model.pvalue %>% 
    filter(Marker != "Test") %>%
    mutate(Del_P_adj = p.adjust(as.numeric(Deltoides), "BH")) %>%
    mutate(Nig_P_adj = p.adjust(as.numeric(Nigra), "BH")) %>%
    mutate(CHR = as.numeric(str_replace(string = Marker, pattern = "(...)(..)_(.*)_(.*)",replacement = "\\2"))) %>%
    mutate(BP = as.numeric(str_replace(string = Marker, pattern = "(...)(..)_(.*)_(.*)", replacement = "\\3")))
  
  # save pvalue file
  write_csv(model.pvalue, paste("../Pvalue_Leaf_DN_noC2/", pheno.name, ".csv", sep = ""))
  
  # Manhattan plot for four variables
  # Deltoides
  Del.man <- model.pvalue %>%
    select(Marker, CHR, BP, Del_P_adj) %>%
    filter(Del_P_adj != "NA")
  manhattan(Del.man, snp = "Marker", p = "Del_P_adj", suggestiveline = -log10(0.05), genomewideline =-log10(0.1), main = paste("Manhattan_Deltoides", pheno.name, sep = "_"))
  dev.copy(png, paste("../Manhattan_Leaf_DN_noC2/Manhattan_Deltoides_", pheno.name, ".png", sep = ""), width = 500, height = 300)
  dev.off()
  # Nigra
  Nig.man <- model.pvalue %>%
    select(Marker, CHR, BP, Nig_P_adj) %>%
    filter(Nig_P_adj != "NA")
  manhattan(Nig.man, snp = "Marker", p = "Nig_P_adj", suggestiveline = -log10(0.05), genomewideline =-log10(0.1), main = paste("Manhattan_Nigra", pheno.name, sep = "_"))
  dev.copy(png, paste("../Manhattan_Leaf_DN_noC2/Manhattan_Nigra_", pheno.name, ".png", sep = ""), width = 500, height = 300)
  dev.off()
}
```



